### PR TITLE
Thoroughly improve import resolution

### LINKED ISSUE #2231: Chunk import pruning
When treeshaking, we retain imports even if those identifiers are tree-shaken out, so you can end up with code like:

```js
import { used, unused } from './chunk';
used();
```

It would be nice to sort out the import pruning issue for chunks that import variables that are tree-shaken specifically within that chunk.

Perhaps the identifier render function can provide some kind of final check whether import bindings are used or not.

### LINKED ISSUE #2576: experimentalPreserveModules opotion doesn't work with import *
- Rollup Version:
- Operating System (or Browser): Debian 9 (WSL)
- Node Version: v11.2.0

### How Do We Reproduce?

https://github.com/kghost/rollup-bug

try build this repo

### Expected Behavior

compile all module to corresponding esm file.

### Actual Behavior

src/m1.js was totally omitted by tree shaking, result output is illegal.

### PR DESCRIPTION
**Pull request by lukastaegert**
This PR contains:
- [x] bugfix
- [ ] feature
- [x] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:
resolves #2576
resolves #2231
 and quite possible a few more

### Description
This is a huge follow-up to #2575. The original intention was to fix handling for dynamic namespaces when preserving modules which lead me down a deep rabbit whole and had me refactoring many core parts of the logic. This is what has been fixed and improved:
* Preserve modules now supports dynamic namespaces which is done not by creating namespace objects but by using a native way of doing a namespace import depending on the output format (i.e. `import * as lib from ...` for esm, `const lib = require(...)` for cjs and so on). Note that namespace freezing will not be performed when preserving modules!
* Shimming missing exports has been thoroughly re-worked and should now work properly for all export formats. Moreover, the shim variable will now always be deconflicted from existing variables of the same name!
* The SystemJS rendering code has been refactored and partially cleaned up to support proper missing export shimming.
* Export tracing has been thoroughly refactored. Most notably, there is no longer any tracing logic at all on chunk level! This was achieved by attaching the exporting module to all variables which also leads to better and more direct imports in some cases.
* The imports of a module are now directly determined during the tree-shaking phase with the awesome benefit that no unnecessary imports are rendered any longer! E.g if `a.js` and `b.js` both import a variable `x` from `c.js` that is only used in one of them, the other will no longer render the import! The effect becomes more pronounced when handling namespace imports. And the new logic is in fact a little simpler than the old one!

### CODE REVIEW

File: test/chunking-form/samples/namespace-object-import/_expected/amd/main2-3b4df12d.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ define(['exports'], function (exports) { 'use strict';
 	});
 
 	exports.a = a;
-	exports.b = b;
```
Comment (guybedford) [inline test/chunking-form/samples/namespace-object-import/_expected/amd/main2-3b4df12d.js]: Wow, did you implement the import pruning!? Very cool. That closes https://github.com/rollup/rollup/issues/2231 if so.

**Comment by guybedford** [comment]
Wondering as well if this has affected https://github.com/rollup/rollup/issues/2556 at all?

**Comment by guybedford** [comment]
Wondering as well if this has affected https://github.com/rollup/rollup/issues/2556 at all?

**Comment by guybedford** [comment]
> This was achieved by attaching the exporting module to all variables which also leads to better and more direct imports in some cases.

One thing to ensure here is that execution ordering is carefully maintained. Previously I remember looking into doing these "direct" imports but found that was causing chunk ordering to be changed in some cases (say chunkA imports X from chunkB which is just reexported from chunkC, it might short-circuit chunkB, and lose the fact that chunkB was supposed to execute).

I don't know if the above applies, but it is worth checking.

**Comment by lukastaegert** [comment]
Thanks for the great reviews! I will check into everything soon. Special thanks for the hint about the execution order. The logic was already there to sort-of maintain execution order by making sure all direct dependencies are added event though nothing was imported from those files but it was broken because the dependency sorting was happening too early. Already added a commit that fixes this.

**Comment by lukastaegert** [comment]
> Wondering as well if this has affected #2556 at all?

Unfortunately not, this is a very separate and probably tricky issue. The question is what the correct SystemJS code should be in such a situation but this should probably be discussed in #2556

**Comment by lukastaegert** [comment]
> Wondering as well if this has affected #2556 at all?

See #2587

**Pull request merged**